File: form_autofill/automation/playwright/src/label-matcher.mjs
const SYNONYMS = {
  father_name: ['father name', "father's name", 'father'],
  mother_name: ['mother name', "mother's name", 'mother'],
  gender: ['gender', 'sex'],
  address: ['address', 'residential address', 'permanent address'],
  dob: ['dob', 'date of birth', 'birth date'],
  district: ['district'],
  full_name: ['full name', 'name', 'applicant name'],
  email: ['email', 'e-mail'],
  phone: ['phone', 'mobile', 'contact'],
  pincode: ['pincode', 'pin code'],
  state: ['state'],
  city: ['city', 'town'],
  aadhaar: ['aadhaar', 'aadhar'],
  pan: ['pan', 'pan card'],
};

function normalize(s) {
  return s.toLowerCase().replace(/[^\w\s]/g, ' ').replace(/\s+/g, ' ').trim();
}

export function matchLabel(raw, minScore = 55) {
  const n = normalize(raw);
  if (!n) return null;
  let bestKey = null;
  let bestScore = 0;
  for (const [key, terms] of Object.entries(SYNONYMS)) {
    for (const term of terms) {
      const t = normalize(term);
      let score = 0;
      if (n === t) score = 100;
      else if (n.includes(t) || t.includes(n)) score = 85;
      else {
        const a = n.split(' ');
        const b = t.split(' ');
        const matches = a.filter((x) => b.some((y) => y === x || y.startsWith(x))).length;
        score = Math.round((matches / a.length) * 100);
      }
      if (score > bestScore) {
        bestScore = score;
        bestKey = key;
      }
    }
  }
  return bestScore >= minScore ? bestKey : null;
}

File: form_autofill/lib/automation/shared/label_matcher.dart
/// Shared fuzzy label matching for all platforms (OCR + DOM + a11y).
class LabelMatcher {
  static const Map<String, List<String>> synonyms = {
    'father_name': [
      'father name',
      'father\'s name',
      'fathers name',
      'father',
      'पिता',
    ],
    'mother_name': [
      'mother name',
      'mother\'s name',
      'mothers name',
      'mother',
      'माता',
    ],
    'gender': ['gender', 'sex', 'लिंग'],
    'address': [
      'address',
      'residential address',
      'permanent address',
      'current address',
      'पता',
    ],
    'dob': ['dob', 'date of birth', 'birth date', 'जन्म तिथि'],
    'district': ['district', 'जिला'],
    'full_name': ['full name', 'name', 'applicant name', 'नाम'],
    'email': ['email', 'e-mail'],
    'phone': ['phone', 'mobile', 'contact', 'mobile number'],
    'pincode': ['pincode', 'pin code', 'postal code'],
    'state': ['state', 'राज्य'],
    'city': ['city', 'town', 'शहर'],
    'aadhaar': ['aadhaar', 'aadhar', 'uid'],
    'pan': ['pan', 'pan card'],
  };

  static String? matchLabel(String rawLabel, {int minScore = 55}) {
    final normalized = _normalize(rawLabel);
    if (normalized.isEmpty) return null;

    String? bestKey;
    var bestScore = 0;

    for (final entry in synonyms.entries) {
      for (final synonym in entry.value) {
        final score = _score(normalized, _normalize(synonym));
        if (score > bestScore) {
          bestScore = score;
          bestKey = entry.key;
        }
      }
    }

    return bestScore >= minScore ? bestKey : null;
  }

  static String _normalize(String input) =>
      input.toLowerCase().replaceAll(RegExp(r'[^\w\s]'), ' ').replaceAll(RegExp(r'\s+'), ' ').trim();

  static int _score(String a, String b) {
    if (a == b) return 100;
    if (a.contains(b) || b.contains(a)) return 85;
    final aTokens = a.split(' ');
    final bTokens = b.split(' ');
    var matches = 0;
    for (final t in aTokens) {
      if (bTokens.any((bt) => bt == t || bt.startsWith(t) || t.startsWith(bt))) {
        matches++;
      }
    }
    if (aTokens.isEmpty) return 0;
    return ((matches / aTokens.length) * 100).round();
  }
}
